### DISCUSSION TITLE
Open Mini browser from command line

### DISCUSSION THREAD

**Discussion by Tpuljak** [eclipse-theia/theia#9986] [Ideas]
It would be great that the Mini browser could be opened from the Terminal.

For example when starting a `react-scripts` project, Chrome open automatically to `localhost:3000`. I am trying to find a way to open the Mini browser instead of Chrome so users can have a browser side by side to their code with hot reload functionality. 

I know that I can input the url in `Preview: Open URL`, but I would like this to be done automatically.

I tried to see how [`open`](https://www.npmjs.com/package/open) works (the npm that `react-scripts` uses) but I didn't have any luck in trying to bind that to Mini browser.

If this can be done, I would happily contribute to the extension to make it work.

**Comment by msujew** [comment]
Well, combining input/output from the backend shell and Theia is always a bit more involved, as everything is running in separate processes. [Gitpod](https://www.gitpod.io) (who were using Theia as their default IDE until April) used a custom Theia extension that send a notification to the frontend anytime a new port opens, which looks like this:

![image](https://user-images.githubusercontent.com/4377073/131329939-1e5e99de-abca-4032-a34e-e23e001a1ca7.png)

I imagine it doesn't make a whole lot of sense to implement this feature for certain scripts/apps. Rather, one should implement it on a container level for all ports.

**Reply by Tpuljak** [reply]
I already have an extension that does something like that, but I was wondering if something could be done to make it a bit more seamless like opening Chrome when React starts.

Is there a way to listen to input/output of the backend shell (on the backend of course)?

**Reply by msujew** [reply]
Sure, the `TerminalServer.create` method returns an `id` that can be fed into the `ProcessManager` to get the appropriate `TerminalProcess`. You can then follow up like in the `TerminalBackendContribution`:

https://github.com/eclipse-theia/theia/blob/02261581de60f5221de6264fab4b011c51e5ab91/packages/terminal/src/node/terminal-backend-contribution.ts#L35-L40

**Reply by Tpuljak** [reply]
Nice. Thank you, I'll see if that's going to help my case or just do what Gitpod does 😄.
